src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5977,6 +5979,105 @@ public static Task<Task> WhenAny(params Task[] tasks)
+        /// <typeparam name="TTask">Specifies the type of the task.</typeparam>
+        /// <remarks>
+        /// This has essentially the same logic as <see cref="TaskFactory.CompleteOnInvokePromise"/>, but optimized
+        /// for two tasks rather than any number. Exactly two tasks has shown to be the most common use-case by far.
+        /// </remarks>
+        private sealed class TwoTaskWhenAnyPromise<TTask> : Task<TTask>, ITaskCompletionAction where TTask : Task
+        {
+            private TTask? _task1, _task2;
+
+            /// <summary>Instantiate the promise and register it with both tasks as a completion action.</summary>
+            public TwoTaskWhenAnyPromise(TTask task1, TTask task2)
+            {
+                Debug.Assert(task1 != null && task2 != null);
+                _task1 = task1;
+                _task2 = task2;
+
+                if (AsyncCausalityTracer.LoggingOn)
+                {
+                    AsyncCausalityTracer.TraceOperationCreation(this, "Task.WhenAny");
+                }
+
+                if (s_asyncDebuggingEnabled)
+                {
+                    AddToActiveTasks(this);
+                }
+
+                task1.AddCompletionAction(this);
+
+                task2.AddCompletionAction(this);
+                if (task1.IsCompleted)
+                {
+                    // If task1 has already completed, Invoke may have tried to remove the continuation from
+                    // each task before task2 added the continuation, in which case it's now referencing the
+                    // already completed continuation.  To deal with that race condition, explicitly check
+                    // and remove the continuation here.
+                    task2.RemoveContinuation(this);
+                }
+            }
+
+            /// <summary>Completes this task when one of the constituent tasks completes.</summary>
+            public void Invoke(Task completingTask)
+            {
+                Task? task1;
+                if ((task1 = Interlocked.Exchange(ref _task1, null)) != null)
+                {
+                    Task? task2 = _task2;
+                    _task2 = null;
+
+                    Debug.Assert(task1 != null && task2 != null);
+                    Debug.Assert(task1.IsCompleted || task2.IsCompleted);
+
+                    if (AsyncCausalityTracer.LoggingOn)
+                    {
+                        AsyncCausalityTracer.TraceOperationRelation(this, CausalityRelation.Choice);
+                        AsyncCausalityTracer.TraceOperationCompletion(this, AsyncCausalityStatus.Completed);
+                    }
+
+                    if (s_asyncDebuggingEnabled)
+                    {
+                        RemoveFromActiveTasks(this);
+                    }
+
+                    if (!task1.IsCompleted)
+                    {
+                        task1.RemoveContinuation(this);
+                    }
```

Comment:
should we make this else and assert task2.IsCompleted?

Reply:
I don't think we need the assert because I added the one above `Debug.Assert(task1.IsCompleted || task2.IsCompleted);`.

But I can add the else.